Use append unique instead of update

aswinsuryan opened the pull request:
When the route-agent after the globalnet controller programs the flow,
it replaces the flows added by globalnet controller. Now append unique adds
the flow and the flows if added by the globalnet controller will get a
higher priority.
Signed-off-by: Aswin Suryanarayanan <aswinsuryan@gmail.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our developer guide: https://submariner.io/development/
2. Ensure you have added the appropriate tests for your PR: https://submariner.io/development/code-review/#test-new-functionality
3. Read the code review guide to ease the review process: https://submariner.io/development/code-review/
4. If the PR is unfinished, mark it as a draft: https://submariner.io/development/code-review/#mark-work-in-progress-prs-as-drafts
5. If you are using CI to debug, use your private fork: https://submariner.io/development/code-review/#use-private-forks-for-debugging-prs-by-running-ci
6. Add labels to the PR as appropriate.

This template is based on the K8s/K8s template:

https://github.com/kubernetes/kubernetes/blob/master/.github/PULL_REQUEST_TEMPLATE.md
-->
Fixes : #1981

aswinsuryan committed:
Use append unique instead of update

When the route-agent after the globalnet controller programs the flow,
it replaces the flows added by globalnet controller. Now append unique adds
the flow and the flows if added by globalnet controller will get a
higher priority.

Signed-off-by: Aswin Suryanarayanan <aswinsuryan@gmail.com>

sridhargaddam reviewed (changes requested):
Can you check if we are deleting the rules when a remote endpoint is removed?

sridhargaddam commented on pkg/routeagent_driver/handlers/ovn/gateway_dataplane.go:
```diff
@@ -155,22 +155,16 @@ func (ovn *Handler) setupForwardingIptables() error {
 	for _, endpoint := range ovn.remoteEndpoints {
 		for _, subnet := range endpoint.Spec.Subnets {
-			rules = append(rules, []string{"-d", subnet, "-j", "ACCEPT"})
+			err := ovn.ipt.AppendUnique("nat", constants.SmPostRoutingChain, []string{"-d", subnet, "-j", "ACCEPT"}...)
```
It looks like we are not deleting the iptable rules when the remote endpoint is deleted.

aswinsuryan committed:
Delete the SmPostRoutingChain rules

Delete SmPostRoutingChain the rules when the endpoint is removed and when
the node transition to non gateway node

Signed-off-by: Aswin Suryanarayanan <aswinsuryan@gmail.com>

sridhargaddam commented on pkg/routeagent_driver/handlers/ovn/handler.go:
```diff
@@ -144,6 +144,12 @@ func (ovn *Handler) RemoteEndpointCreated(endpoint *submV1.Endpoint) error {
 	if ovn.isGateway {
+		for _, subnet := range endpoint.Spec.Subnets {
+			if err = ovn.addNoMasqueradeIPTables(subnet); err != nil {
+				return errors.Wrap(err, "error adding no-masquerade rules")
```
Please include the subnet (or the rule) for which the error occurred.

aswinsuryan replied:
Done.

sridhargaddam commented on pkg/routeagent_driver/handlers/ovn/handler.go:
```diff
@@ -186,6 +192,12 @@ func (ovn *Handler) RemoteEndpointRemoved(endpoint *submV1.Endpoint) error {
 	if ovn.isGateway {
+		for _, subnet := range endpoint.Spec.Subnets {
+			if err = ovn.removeNoMasqueradeIPTables(subnet); err != nil {
+				return errors.Wrap(err, "error removing no-masquerade rules")
```
same

aswinsuryan replied:
Done.

sridhargaddam commented on pkg/routeagent_driver/handlers/ovn/handler.go:
```diff
@@ -197,6 +209,13 @@ func (ovn *Handler) TransitionToNonGateway() error {
+	for _, endpoint := range ovn.remoteEndpoints {
+		for _, subnet := range endpoint.Spec.Subnets {
+			if err := ovn.removeNoMasqueradeIPTables(subnet); err != nil {
+				return errors.Wrap(err, "error removing no-masquerade rules")
```
same

aswinsuryan replied:
Done

sridhargaddam commented on pkg/routeagent_driver/handlers/ovn/handler.go:
```diff
@@ -207,5 +226,13 @@ func (ovn *Handler) TransitionToGateway() error {
+	for _, endpoint := range ovn.remoteEndpoints {
+		for _, subnet := range endpoint.Spec.Subnets {
+			if err := ovn.addNoMasqueradeIPTables(subnet); err != nil {
+				return errors.Wrap(err, "error adding no-masquerade rules")
```
same

sridhargaddam commented on pkg/routeagent_driver/handlers/ovn/handler.go:
```diff
@@ -144,6 +144,12 @@ func (ovn *Handler) RemoteEndpointCreated(endpoint *submV1.Endpoint) error {
 	}
 
 	if ovn.isGateway {
+		for _, subnet := range endpoint.Spec.Subnets {
```
We will still have to handle the out of order events which can trigger race conditions. Basically, we have to look at the lastProcessedTime and process/ignore the event similar to how we are doing in kubeproxy handler - https://github.com/submariner-io/submariner/blob/0c3e043a374c6a793c564a9d98fa8fb8990b310e/pkg/routeagent_driver/handlers/kubeproxy/endpoint_handler.go#L108

But anyway, we can use a separate PR for that.

aswinsuryan replied:
sure will do as a follow up pr

aswinsuryan committed:
Add subnet details in logging

Signed-off-by: Aswin Suryanarayanan <aswinsuryan@gmail.com>

aswinsuryan committed:
Merge branch 'devel' into ovn-globalnet

sridhargaddam reviewed (approved):
Thanks for addressing the comments @aswinsuryan
